Add groupBy for vectors

alexkalderimis opened the pull request:
This adds `groupBy` for generic vectors, returning a list of
subvectors.

This is similar to #143, aiming to resolve #192 

Some of the comments on #143 have been applied here.

Input on testing would be welcome

Shimuuar commented:
Looks reasonable to me. Could you add specialized version to Data.Vector, etc and add some haddocks?

alexkalderimis commented:
Thanks for taking a look @Shimuuar - I'll ping you once this has been updated.

konsumlamm commented on vector/src/Data/Vector.hs:
```diff
@@ -1410,6 +1410,39 @@ break :: (a -> Bool) -> Vector a -> (Vector a, Vector a)
+-- >>> groupBy (\a b -> isUpper a == isUpper b) (fromList "Mississippi River")
+-- ["M", "ississippi ", "R", "iver"]
+--
+-- See 'Data.List.groupBy'
```
```suggestion
-- See also 'Data.List.groupBy'.
```

konsumlamm commented on vector/src/Data/Vector.hs:
```diff
@@ -1410,6 +1410,39 @@ break :: (a -> Bool) -> Vector a -> (Vector a, Vector a)
+-- The concatenation of this list of slices is equal to the argument vector,
+-- and each slice contains only equal elements.
+--
+-- This is the overloaded version of 'groupBy'.
```
```suggestion
-- This is equivalent to 'groupBy' (==).
```

konsumlamm commented on vector/src/Data/Vector.hs:
```diff
@@ -1410,6 +1410,39 @@ break :: (a -> Bool) -> Vector a -> (Vector a, Vector a)
+-- >>> group (fromList "Mississippi")
+-- ["M","i","ss","i","ss","i","pp","i"]
+--
+-- See 'Data.List.group'
```
```suggestion
-- See also 'Data.List.group'.
```

konsumlamm commented on vector/src/Data/Vector/Generic.hs:
```diff
@@ -1552,6 +1552,44 @@ break f xs = case findIndex f xs of
+-- >>> groupBy (\a b -> isUpper a == isUpper b) (fromList "Mississippi River")
+-- ["M", "ississippi ", "R", "iver"]
+--
+-- See 'Data.List.groupBy'
```
```suggestion
-- See also 'Data.List.groupBy'.
```

konsumlamm commented on vector/src/Data/Vector/Generic.hs:
```diff
@@ -1552,6 +1552,44 @@ break f xs = case findIndex f xs of
+-- The concatenation of this list of slices is equal to the argument vector,
+-- and each slice contains only equal elements.
+--
+-- This is the overloaded version of 'groupBy'.
```
```suggestion
-- This is equivalent to 'groupBy (==)'.
```

konsumlamm commented on vector/src/Data/Vector/Generic.hs:
```diff
@@ -1552,6 +1552,44 @@ break f xs = case findIndex f xs of
+-- >>> group (fromList "Mississippi")
+-- ["M","i","ss","i","ss","i","pp","i"]
+--
+-- See 'Data.List.group'
```
```suggestion
-- See also 'Data.List.group'.
```

konsumlamm commented on vector/src/Data/Vector/Primitive.hs:
```diff
@@ -1169,6 +1169,29 @@ break :: Prim a => (a -> Bool) -> Vector a -> (Vector a, Vector a)
+-- The concatenation of this list of slices is equal to the argument vector,
+-- and each slice contains only equal elements.
+--
+-- This is the overloaded version of 'groupBy'.
```
```suggestion
-- This is equivalent to 'groupBy (==)'.
```

konsumlamm commented on vector/src/Data/Vector/Storable.hs:
```diff
@@ -1190,6 +1190,29 @@ break :: Storable a => (a -> Bool) -> Vector a -> (Vector a, Vector a)
+-- The concatenation of this list of slices is equal to the argument vector,
+-- and each slice contains only equal elements.
+--
+-- This is the overloaded version of 'groupBy'.
```
```suggestion
-- This is equivalent to 'groupBy (==)'.
```

konsumlamm commented on vector/src/Data/Vector/Unboxed.hs:
```diff
@@ -1182,6 +1182,27 @@ break :: Unbox a => (a -> Bool) -> Vector a -> (Vector a, Vector a)
+-- The concatenation of this list of slices is equal to the argument vector,
+-- and each slice contains only equal elements.
+--
+-- This is the overloaded version of 'groupBy'.
```
```suggestion
-- This is equivalent to 'groupBy (==)'.
```

alexkalderimis commented:
@Shimuuar - the haddocks and specialized versions should be ready for your review now

Shimuuar commented:
Thank you! PR look great.

 Doctests fails because of extra whitespace. And could also mentiaion that function does not fuse, and resulting vectors are slices of original vector

alexkalderimis commented:
> Thank you! PR look great.

Thanks
 
> Doctests fails because of extra whitespace. 

D'oh - easy fix at least

> And could also mentiaion that function does not fuse

Can do - thought this was not necessary since the result of the function is a list, not a Vector

> and resulting vectors are slices of original vector

The haddocks say: "Split a vector into a list of slices." - is this enough?

alexkalderimis commented:
Back to you, @Shimuuar

Shimuuar commented:
Still one test failure. (It's really annoying that doctests only work on 8.6 & 8.8) otherwise good to go. Also could you add changelog entry while you at it

alexkalderimis commented:
> Still one test failure. (It's really annoying that doctests only work on 8.6 & 8.8) otherwise good to go. Also could you add changelog entry while you at it

It is a little annoying, especially since when I use 8.8.4 locally and run the test suite, I get errors like:

```
src/Data/Vector.hs:978: failure in expression `import qualified Data.Vector as V'
expected: 
 but got: 
          ^
          <no location info>: error:
              Could not find module ‘Data.Vector’
              Perhaps you meant Data.Functor (from base-4.13.0.0)
```

Tips on reproducing doctests locally would certainly, be appreciated, but I notice that the checks are running now on push so that is not a big deal.

It looks like I have fixed the remaining issues, so I have added the changelog entry, and squashed the commits for a cleaner history. Back to you @Shimuuar
